framework/src/preconditioners/DualMortarPreconditioner.C

```diff
@@ -0,0 +1,588 @@
+
+      std::vector<dof_id_type> vec_dofs(dofs.begin(), dofs.end());
+      _dof_sets_interior[vn].insert(make_pair(sd, vec_dofs));
+    }
+  }
+}
+
+void
+DualMortarPreconditioner::getDofVarInterface()
+{
+  _dof_sets_primary.resize(_n_vars);
+  _dof_sets_secondary.resize(_n_vars);
+  for (unsigned int vn = 0; vn < _n_vars; vn++)
+  {
+    // loop over boundary nodes
+    ConstBndNodeRange & range = * _mesh->getBoundaryNodeRange();
+    std::vector<dof_id_type> di;
+    for (const auto & bnode : range)
+    {
+      const Node * node_bdry = bnode->_node;
+      BoundaryID boundary_id = bnode->_bnd_id;
+
+      if (boundary_id == _secondary_boundary)
+      {
+          _dofmap->dof_indices(node_bdry, di, vn);
+          for (auto it = di.begin(); it!=di.end(); ++it)
+            if(std::find(_dof_sets_secondary[vn].begin(), _dof_sets_secondary[vn].end(), *it)==_dof_sets_secondary[vn].end())
+              _dof_sets_secondary[vn].push_back(*it);
+      }
+
+      if (boundary_id == _primary_boundary)
+      {
+        _dofmap->dof_indices(node_bdry, di, vn);
+        for (auto it = di.begin(); it!=di.end(); ++it)
+          if(std::find(_dof_sets_primary[vn].begin(), _dof_sets_primary[vn].end(), *it)==_dof_sets_primary[vn].end())
+            _dof_sets_primary[vn].push_back(*it);
+      }
+    }
+  }
+}
+
+void
+DualMortarPreconditioner::printDofSets()
+{
+  /// Dofs in the interior
+  for (auto it1 = _dof_sets_interior.begin(); it1 != _dof_sets_interior.end(); it1++)
+  {
+    std::cout << "variable "<<std::endl;
+    for (auto it2 = it1->begin(); it2 != it1->end(); it2++)
+    {
+      std::cout << " \tdofs in subdomain " << it2->first<<": ";
+      for (auto it3 = it2->second.begin(); it3 != it2->second.end(); it3++)
+        std::cout <<*it3 << " ";
+      std::cout <<"\n";
+    }
+  }
+
+  // primary
+  std::cout<<"primary dofs:\n";
+  for (auto it1=_dof_sets_primary.begin(); it1!=_dof_sets_primary.end(); it1++)
+  {
+    std::cout<< "\tvariable :";
+    for (auto it2=it1->begin(); it2!=it1->end(); it2++)
+      std::cout<<*it2<<" ";
+    std::cout<<std::endl;
+  }
+
+  // secondary
+  std::cout<<"secondary dofs:\n";
+  for (auto it1=_dof_sets_secondary.begin(); it1!=_dof_sets_secondary.end(); it1++)
+  {
+    std::cout<< "\tvariable :";
+    for (auto it2=it1->begin(); it2!=it1->end(); it2++)
+      std::cout<<*it2<<" ";
+    std::cout<<std::endl;
+  }
+}
+
+void
+DualMortarPreconditioner::condenseSystem()
+{
+  std::vector<dof_id_type> u1c=_dof_sets_primary[0];
+  std::vector<dof_id_type> u2c=_dof_sets_secondary[0];
+
+  std::vector<dof_id_type> lm=_dof_sets_secondary[1];
+
+  std::vector<dof_id_type> u1i=_dof_sets_interior[0][_primary_subdomain];
+  std::vector<dof_id_type> u2i=_dof_sets_interior[0][_secondary_subdomain];
+
+  _matrix->create_submatrix(* _D, lm, u2c);
+  _D->get_transpose(*_D);
+  _matrix->create_submatrix(* _M, lm, u1c);
+  _matrix->create_submatrix(* _MDinv, u1c, lm);
+  _M->get_transpose(*_M);
+  _M->close();
```

Comment:
No close

Reply:
> No close

Why is that? I thought I should close the matrix every time I am done with changing the values?